spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/MappingSpannerReadConverter.java

```diff
@@ -73,163 +78,201 @@
+			}
+			else {
+				result = getPotentiallyConvertedIterable(columnName, innerType);
+			}
+			return result;
 		}
 
-		if (!valueSet) {
-			for (Class sourceType : this.readIterableMapping.keySet()) {
+		private Object getPotentiallyConvertedIterable(String columnName, Class innerType) {
+			for (Class sourceType : readIterableMapping.keySet()) {
 				if (canConvert(sourceType, innerType)) {
-					List iterableValue = readIterableMapping.get(sourceType).apply(struct,
-							colName);
-					accessor.setProperty(spannerPersistentProperty, ConversionUtils
-							.convertIterable(iterableValue, innerType, this));
-					valueSet = true;
-					break;
+					List iterableValue = readIterableMapping.get(sourceType).apply(this.struct, columnName);
+					return ConversionUtils
+							.convertIterable(iterableValue, innerType, MappingSpannerReadConverter.this);
 				}
 			}
+			return null;
 		}
 
-		if (!valueSet && struct.getColumnType(colName).getArrayElementType().getCode() == Type.Code.STRUCT) {
-			List<Struct> iterableValue = struct.getStructList(colName);
-			Iterable convertedIterableValue = (Iterable) StreamSupport.stream(iterableValue.spliterator(), false)
-					.map(item -> read(innerType, item))
-					.collect(Collectors.toList());
-			accessor.setProperty(spannerPersistentProperty, convertedIterableValue);
-			valueSet = true;
+		private boolean isSpannerSupportedIterableType(Class innerType) {
+			return readIterableMapping.containsKey(innerType);
 		}
 
-		return valueSet;
-	}
+		private boolean isArrayOfStruct(String columnName) {
+			return this.struct.getColumnType(columnName).getArrayElementType().getCode() == Type.Code.STRUCT;
+		}
 
-	private static <R> R instantiate(Class<R> type) {
-		R object;
-		try {
-			Constructor<R> constructor = type.getDeclaredConstructor();
-			constructor.setAccessible(true);
-			object = constructor.newInstance();
+		private Object getSingle(SpannerPersistentProperty spannerPersistentProperty) {
+			String columnName = spannerPersistentProperty.getColumnName();
+			Class sourceType = ConversionUtils.SPANNER_COLUMN_TYPE_TO_JAVA_TYPE_MAPPING
+					.get(this.struct.getColumnType(columnName));
+
+			Object result = null;
+			Class targetType = spannerPersistentProperty.getType();
+			if (sourceType != null && canConvert(sourceType, targetType)) {
+				BiFunction readFunction = singleItemReadMethodMapping.get(ConversionUtils.boxIfNeeded(sourceType));
+				if (readFunction != null) {
+					result = convert(readFunction.apply(this.struct, columnName), targetType);
+				}
+			}
+			return result;
 		}
-		catch (ReflectiveOperationException e) {
-			throw new SpannerDataException(
-					"Unable to create a new instance of entity using default constructor.",
-					e);
+
+		public boolean isNull(String columnName) {
+			return this.struct.isNull(columnName);
 		}
-		return object;
-	}
 
+		public boolean hasColumn(String columnName) {
+			return this.struct.getType().getStructFields().stream()
```

Comment:
Using Java 8 `Stream` in conversion/mapping code paths has a negative impact on performance (in MongoDB we had a 100% performance impact, mapping took twice as long using Java 8 `Stream`).

Reply:
Thank you, I'll change that, and check for other parts too.